### ISSUE TITLE
emcmake can't find pthread

### ISSUE THREAD

**Issue by LeenHawk** [emscripten-core/emscripten#18948]
Please include the following in your bug report:

**Version of emscripten/emsdk:**
```shell
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.32 (eab98adf462c39f3c31d348331c4830bcaa36949)
clang version 17.0.0 (https://github.com/llvm/llvm-project df82394e7a2d06506718cafa347bf7827c79fc4f)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /home/leen/tools/emsdk/upstream/bin
```
**Failing command line in full:**
If this is compile or link-time failure please include the full failing command
along with its entire output.
```shell
git clone https://github.com/facebook/zstd.git
mkdir zstd/build/cmake/builddir && cd zstd/build/cmake/builddir
emcmake cmake .. -DCMAKE_CXX_FLAGS="-pthread" -DCMAKE_C_FLAGS="-pthread"
```
output:
```shell
configure: cmake .. -DCMAKE_CXX_FLAGS=-pthread -DCMAKE_C_FLAGS=-pthread -DCMAKE_TOOLCHAIN_FILE=/home/saltyfish/tools/emsdk/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake -DCMAKE_CROSSCOMPILING_EMULATOR=/home/saltyfish/tools/emsdk/node/14.18.2_64bit/bin/node;--experimental-wasm-bulk-memory;--experimental-wasm-threads
-- The ASM leen identification is unknown
-- Found assembler: /home/saltyfish/tools/emsdk/upstream/emscripten/emcc
-- Warning: Did not find file leen/-ASM
-- ZSTD VERSION: 1.5.4
-- Setting build type to 'Release' as none was specified.
-- Performing Test C_FLAG_WALL
-- Performing Test C_FLAG_WALL - Success
-- Performing Test CXX_FLAG_WALL
-- Performing Test CXX_FLAG_WALL - Success
-- Performing Test C_FLAG_WEXTRA
-- Performing Test C_FLAG_WEXTRA - Success
-- Performing Test CXX_FLAG_WEXTRA
-- Performing Test CXX_FLAG_WEXTRA - Success
-- Performing Test C_FLAG_WUNDEF
-- Performing Test C_FLAG_WUNDEF - Success
-- Performing Test CXX_FLAG_WUNDEF
-- Performing Test CXX_FLAG_WUNDEF - Success
-- Performing Test C_FLAG_WSHADOW
-- Performing Test C_FLAG_WSHADOW - Success
-- Performing Test CXX_FLAG_WSHADOW
-- Performing Test CXX_FLAG_WSHADOW - Success
-- Performing Test C_FLAG_WCAST_ALIGN
-- Performing Test C_FLAG_WCAST_ALIGN - Success
-- Performing Test CXX_FLAG_WCAST_ALIGN
-- Performing Test CXX_FLAG_WCAST_ALIGN - Success
-- Performing Test C_FLAG_WCAST_QUAL
-- Performing Test C_FLAG_WCAST_QUAL - Success
-- Performing Test CXX_FLAG_WCAST_QUAL
-- Performing Test CXX_FLAG_WCAST_QUAL - Success
-- Performing Test C_FLAG_WSTRICT_PROTOTYPES
-- Performing Test C_FLAG_WSTRICT_PROTOTYPES - Success
-- Performing Test LD_FLAG_Z_NOEXECSTACK
-- Performing Test LD_FLAG_Z_NOEXECSTACK - Success
-- Performing Test C_FLAG_QUNUSED_ARGUMENTS
-- Performing Test C_FLAG_QUNUSED_ARGUMENTS - Failed
-- Performing Test CXX_FLAG_QUNUSED_ARGUMENTS
-- Performing Test CXX_FLAG_QUNUSED_ARGUMENTS - Failed
-- Performing Test C_FLAG_WA_NOEXECSTACK
-- Performing Test C_FLAG_WA_NOEXECSTACK - Failed
-- Performing Test CXX_FLAG_WA_NOEXECSTACK
-- Performing Test CXX_FLAG_WA_NOEXECSTACK - Failed
-- CMAKE_INSTALL_PREFIX: /home/saltyfish/tools/emsdk/upstream/emscripten/cache/sysroot
-- CMAKE_INSTALL_LIBDIR: lib
-- ZSTD_LEGACY_SUPPORT defined!
-- ZSTD_MULTITHREAD_SUPPORT is enabled
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if leen accepts -pthread
-- Check if leen accepts -pthread - no
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
CMake Error at /usr/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Threads (missing: Threads_FOUND)
Call Stack (most recent call first):
  /usr/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.25/Modules/FindThreads.cmake:228 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:132 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/saltyfish/tools/third_party/zstd/build/cmake/builddir/CMakeFiles/CMakeOutput.log".
See also "/home/saltyfish/tools/third_party/zstd/build/cmake/builddir/CMakeFiles/CMakeError.log".
emcmake: error: 'cmake .. -DCMAKE_CXX_FLAGS=-pthread -DCMAKE_C_FLAGS=-pthread -DCMAKE_TOOLCHAIN_FILE=/home/saltyfish/tools/emsdk/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake -DCMAKE_CROSSCOMPILING_EMULATOR=/home/saltyfish/tools/emsdk/node/14.18.2_64bit/bin/node;--experimental-wasm-bulk-memory;--experimental-wasm-threads' failed (returned 1)
```
Is there a way to enable pthread in  emscripten using cmake command parameter?

**Add Files:**
[CMakeError.log](https://github.com/emscripten-core/emscripten/files/10963372/CMakeError.log)
[CMakeOutput.log](https://github.com/emscripten-core/emscripten/files/10963373/CMakeOutput.log)

**Comment by sbc100** [comment]
Can you include the `CMakeOutput.log` amd `CMakeError.log` files?

**Comment by sbc100** [comment]
Looks like the problem is that `cmake/CMakeModules/AddZstdCompilationFlags.cmake` does EnableCompilerFlag("-z noexecstack") which results in:

```
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_aed62.dir/link.txt --verbose=1
/usr/local/google/home/sbc/dev/wasm/emscripten/emcc  -Wall -Wextra -Wundef -Wshadow -Wcast-align -Wcast-qual -Wstrict-prototypes   -z noexecstack  @CMakeFiles/cmTC_aed62.dir/objects1 -o cmTC_aed62.js-
emcc: error: noexecstack: No such file or directory ("noexecstack" was expected to be an input file, based on the commandline arguments provided)
gmake[1]: *** [CMakeFiles/cmTC_aed62.dir/build.make:101: cmTC_aed62.js] Error 1   
```

emscripten doesn't support that `-z` flag, so you would want to modify AddZstdCompilationFlags.cmake to add it conditionally.

**Comment by sbc100** [comment]
Actually unknown `-z` flags are only supposed to be a linker warning.  I noticed we were not correctly passing `-z` flags through to the linker.  I created a fix in #18956

**Comment by LeenHawk** [comment]
> Can you include the `CMakeOutput.log` amd `CMakeError.log` files?

I included now.
